File: android/app/src/main/java/com/emwaver/emwaverandroidapp/TransportDeviceSessionRegistry.java
/*
 * EMWaver
 * Copyright (c) 2026 Luís Marnoto
 * SPDX-License-Identifier: Apache-2.0
 */

package com.emwaver.emwaverandroidapp;

import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

final class TransportDeviceSessionRegistry {
    private final Map<String, TransportDeviceSession> sessionsByDeviceId = new HashMap<>();
    private TransportDeviceSession activeSession = new DeviceBufferSession();

    synchronized TransportDeviceSession active() {
        return activeSession;
    }

    synchronized TransportDeviceSession session(String deviceId) {
        String key = normalize(deviceId);
        TransportDeviceSession session = sessionsByDeviceId.get(key.toLowerCase(Locale.US));
        if (session == null) {
            session = new DeviceBufferSession(key);
            sessionsByDeviceId.put(key.toLowerCase(Locale.US), session);
        }
        return session;
    }

    synchronized TransportDeviceSession select(String deviceId, boolean resetSession) {
        TransportDeviceSession session = session(deviceId);
        activeSession = session;
        if (resetSession) {
            session.clearAll();
        }
        return session;
    }

    private static String normalize(String deviceId) {
        return deviceId == null || deviceId.trim().isEmpty() ? "active" : deviceId.trim();
    }
}

File: ios/EMWaver/Managers/TransportDeviceSessionRegistry.swift
/*
 * EMWaver
 * Copyright (c) 2026 Luís Marnoto
 * SPDX-License-Identifier: Apache-2.0
 */

import Foundation

final class TransportDeviceSessionRegistry {
    private var sessionsByDeviceId: [String: TransportDeviceSession] = [:]
    private var activeSession: TransportDeviceSession = DeviceBufferSession()

    var active: TransportDeviceSession {
        activeSession
    }

    func session(deviceId: String) -> TransportDeviceSession {
        let key = Self.normalize(deviceId)
        let lookupKey = key.lowercased()
        if let session = sessionsByDeviceId[lookupKey] {
            return session
        }

        let session = DeviceBufferSession()
        sessionsByDeviceId[lookupKey] = session
        return session
    }

    @discardableResult
    func select(deviceId: String, resetSession: Bool) -> TransportDeviceSession {
        let session = session(deviceId: deviceId)
        activeSession = session
        if resetSession {
            session.clearAll()
        }
        return session
    }

    private static func normalize(_ deviceId: String) -> String {
        let key = deviceId.trimmingCharacters(in: .whitespacesAndNewlines)
        return key.isEmpty ? "active" : key
    }
}
